### ISSUE TITLE
Feature request: Inception v3 MetaGraph

### ISSUE THREAD

**Issue by Hvass-Labs** [tensorflow/tensorflow#5036]
Would it be possible for the TensorFlow developers to put a tar-ball online with the Inception v3 saved as a MetaGraph? I can't find it anywhere.

I'm currently using the following tar-ball with a frozen graph for Inception v3:

http://download.tensorflow.org/models/image/imagenet/inception-2015-12-05.tgz

The problem is that I cannot continue training that graph because it is frozen, so all the variables have been converted to constants before it was saved. I can't find a way to convert the constants back to variables so I don't think that is possible. (There are also some deprecation warnings regarding BatchNormWithGlobalNormalization so it will presumably stop working at some point in the future).

After searching for a solution for days, I found that you have released a newer checkpoint-file for Inception v3:

http://download.tensorflow.org/models/image/imagenet/inception-v3-2016-03-01.tar.gz

I downloaded it but it's only the checkpoint-file, not the graph-definition. So in my Python code I would apparently have to create the Inception graph using this function first:

https://github.com/tensorflow/models/blob/master/inception/inception/slim/inception_model.py#L52

But this apparently requires building TensorFlow from source, as far as I could understand from the README. There's also several options for using the function and it apparently has to be wrapped in arg_scopes and what-not:

https://github.com/tensorflow/models/blob/master/inception/inception/inception_model.py#L76-87

Would it be possible to update the above tar-ball (dated 2016-03-01) so it also contains the MetaGraph-files, so I can load it more easily and use it in my own Python program? I have another data-set so I replace the softmax-layer of the Inception-graph, and I also want to continue optimizing the rest of the variables of the Inception-graph.

Please also consider including a small example program in the tar-ball (or a link to some python-code), as it would make it a lot easier for everyone who wants to use it. Or at least make a list of all the relevant tensor-names (input, output, etc.)

Thanks!

**Comment by drpngx** [comment]
@martinwicke any reason why we don't ship the graphs?

**Comment by martinwicke** [comment]
No reason other than finiteness of resources. sguada@ may have
more details, but I believe that the code in tensorflow/models produces the
proper graph, and could serve as the requested code example.

I don't think building from source is a requirement. I believe the (Python)
files would work fine with an installed binary version of tensorflow. You
may have to check out tensorflow if the files are not contained in the pip
package.

**Comment by ryszard** [comment]
@martinwicke I am trying to do something very similar to what @Hvass-Labs is attempting.

The inception code in tensorflow/models is doing rather advanced stuff (using multiple GPUs, etc), which goes over the head of a beginner (like me), and it seems rather tightly coupled. What is more, it uses a (custom) version of slim, which not only makes the code look different than mainstream TF code, but also forces you to use blaze (not super pleasant if you aren't already using blaze for everything) – you cannot just drop one file into your repo and count it will work.

(Another source of confusion for me is that there's two versions of the slim inception: one in tensorflow/models, and one in https://github.com/tensorflow/models/tree/master/inception/inception/slim, which are very similar, but slightly different). 

The inception related code that's much more beginner friendly (and much less Google idiosyncratic) is the one used in  https://www.tensorflow.org/versions/r0.9/how_tos/image_retraining/index.html. However, adapting it to training the full network has proven quite daunting (issues mentioned by @Hvass-Labs, like being frozen, etc).

So, getting the Metagraph would actually be very, very helpful. Some non-slim Python code would be even better, but I understand that the rewrite would be far from trivial.

**Comment by Hvass-Labs** [comment]
Thanks for the input everyone.

I am using the Inception model in my tutorials and I try to keep them fairly simple. This means limiting each tutorial to only one or a few topics. The official tutorials have 10 topics in each tutorial which is an extremely steep learning curve for beginners. I've put the Inception model inside a Python module so it is encapsulated and can easily be loaded from different tutorials:

https://github.com/Hvass-Labs/TensorFlow-Tutorials/blob/master/inception.py

By far the easiest thing would be to replace a few lines in my inception.py with the following:

- Download a newer tar-ball.
- A few lines of code for loading the MetaGraph and Checkpoint.

This would be an elegant solution that I could easily explain in a tutorial.

For the end-user, there really is no reason to build the Inception graph all over again. It is quite messy and you can't make any changes to it, because then you cannot load the checkpoint anymore. Even tiny changes that are logically identical will break the checkpoint.

As far as I understand, the MetaGraph is already saved when you save the checkpoint. So it should be fairly trivial to update the tar-ball. Although I do understand that if you guys are under a lot of stress then there's a million 'fairly trivial' things that needs to be done.

Hopefully you'll see the usefulness of updating the tar-ball as it might help thousands of people who will be watching the tutorials in the following years. I've been stuck for 2 weeks now trying to obtain the MetaGraph so I'm hoping I can get it soon.

Cheers!

**Comment by jurastm** [comment]
Same about me. I need to use tar-ball in my project

**Comment by paolodoors** [comment]
Hvass-Labs's tutorials are really great. It would be very helpful to have the MetaGraph available.

**Comment by jfhauris** [comment]
Hvass-Labs's tutorials are outstanding.  Please provide the MetaGraph etc that he is requesting.  Thanks, Jon

**Comment by chidochipotle** [comment]
Hvass-Labs's tutorials are amazing. Would be great if you could implement the MetaGraph so he can make tutorial 10. Thanks, Daniel

**Comment by Arora90** [comment]
Hvass-Labs's tutorials are very helpful. Would be great if you could implement the MetaGraph.

**Comment by souterjk** [comment]
Bump for asking for a meta graph

**Comment by Caspain** [comment]
Hvass-Labs's tutorials are outstanding. Please provide the MetaGraph etc that he is requesting.Everyone deeply appreciates his lessons. Kindly support his request
Thank You.

**Comment by prudhvinandan** [comment]
Hvass-Labs's tutorials are outstanding. Please provide the MetaGraph etc that he is requesting.

**Comment by karl596** [comment]
Hvass-Labs的教程非常出色。请提供他要求的MetaGraph等。

**Comment by Caspain** [comment]
Please support. Give the required files

**Comment by jurastm** [comment]
@karl596 tensorflow 刚出了 v.1.0. 
最新版应该有metagraph

**Comment by vshkola** [comment]
+1 for the really necessary feature.

**Comment by Caspain** [comment]
Please support this feature request. So the community surrounding tensor
can mature further,thank You.

On Tue, Mar 14, 2017 at 10:52 AM, Vladyslav Shkola <notifications@github.com

**Comment by Caspain** [comment]
Hvass-Labs's tutorials are outstanding. Please provide the MetaGraph etc
that he is requesting. Thus extending the usage of this great and best
machine learning library.

**Comment by Caspain** [comment]
Hvass-Labs's tutorials are outstanding. Please provide the MetaGraph etc
that he is requesting

**Comment by Caspain** [comment]
Hvass-Labs's tutorials are outstanding. Please provide the MetaGraph etc
that he is requesting

**Comment by Caspain** [comment]
Hvass-Labs's tutorials are outstanding. Please provide the MetaGraph etc
that he is requesting

**Comment by Caspain** [comment]
Hvass-Labs's tutorials are outstanding. Please provide the MetaGraph etc
that he is requesting

**Comment by Caspain** [comment]
Hvass-Labs's tutorials are outstanding. Please provide the MetaGraph etc
that he is requesting

**Comment by KoSv** [comment]
please support!

**Comment by Caspain** [comment]
please support this feature request.thank you!

**Comment by ElchinValiyev** [comment]
+1

**Comment by ljubisap** [comment]
+1

**Comment by yamashi** [comment]
+1

**Comment by merryHunter** [comment]
Please, support! +1

**Comment by attakorn1** [comment]
+1

**Comment by SinghRanjodh** [comment]
+1

**Comment by shrikondra** [comment]
+1

**Comment by mueller91** [comment]
+1

**Comment by alwynmathew** [comment]
+1

**Comment by YaroslavSchubert** [comment]
+1

**Comment by tanayjha** [comment]
+1

**Comment by gongenhao** [comment]
+1

**Comment by ram1993** [comment]
+1

**Comment by bzhong2** [comment]
+1

**Comment by Vijayabhaskar96** [comment]
+1

**Comment by trojana** [comment]
+1

**Comment by hzqjyyx** [comment]
+1

**Comment by kencano** [comment]
+1

**Comment by wgong** [comment]
Hvass-Labs tensorflow tutorial is awesome, the best I can find, please support his pull request, so we can learn more about tensorflow from Hvass-Labs, Thank tensorflow team

**Comment by RobertCollyer** [comment]
+1

**Comment by thorgeirk11** [comment]
+1

**Comment by gnavihs** [comment]
+1

**Comment by jematos92** [comment]
+1

**Comment by sguada** [comment]
Please use [export_inference_graph](https://github.com/tensorflow/models/blob/master/slim/export_inference_graph.py) to get the graph (see [Docs](https://github.com/tensorflow/models/blob/master/slim/README.md#Export))

**Comment by mrStasSmirnoff** [comment]
+1

**Comment by ajinkyaT** [comment]
+1

**Comment by LeeKyungMoon** [comment]
@sguada The link you added is not working now.

**Comment by sguada** [comment]
Fixed the links

**Comment by grankt** [comment]
+1

**Comment by AymanASamyM** [comment]
+1

**Comment by letuantang** [comment]
+1

**Comment by luongthomas** [comment]
+1

**Comment by letuantang** [comment]
+1

**Comment by omarg42** [comment]
+1

**Comment by nimral** [comment]
+1

**Comment by AbhiRepository** [comment]
+1
